Implement support for markup in wxDataViewCtrl

vadz opened the pull request:
Add `wxDataViewTextRenderer::EnableMarkup()` for finer-grained control over items appearance.

vadz committed:
Add wxDataViewTextRenderer::EnableMarkup()

Implement the new method to all the implementations (generic, GTK, OS X), show
it in the sample and update the documentation.

vadz committed:
Don't pass nor use wxCONTROL_FOCUSED in wxRenderer::DrawItemText()

It doesn't make much sense to always use wxCONTROL_FOCUSED in wxDataViewCtrl
code whenever wxDATAVIEW_CELL_SELECTED state bit is set, wxCONTROL_SELECTED
seems to be enough and is with the native MSW wxRenderer implementation which
doesn't evne test for wxCONTROL_FOCUSED anyhow, while the generic one did and
didn't use the correct colour unless it was set.

No real changes, but this will make upcoming commits more straightforward and
already simplifies the code a little bit.

vadz committed:
Add wxMarkupText::RenderItemText()

Add new method allowing to use wxRendererNative::DrawItemText() for actually
rendering the text instead of wxDC::DrawLabel().

This will be used for markup support in (generic) wxDataViewCtrl.

discnl commented:
I only skimmed through the changes and have some minor typo/style remarks:

In the commit message of 0c7424f22ebd0385d7dc98b720aabcf8343e4e59 `evne` should be `even`.

In the changes of the other two commits I think "\* " should be replaced by " \*", e.g. `wxWindow* win` becomes `wxWindow *win`. Unless it's OK to introduce that style in new wx code?

discnl commented on interface/wx/dataview.h:
```diff
@@ -1966,6 +1966,49 @@ class wxDataViewTextRenderer : public wxDataViewRenderer
                            int align = wxDVR_DEFAULT_ALIGNMENT );
+
+    /**
+        Enable interpretation of markup in the cells data.
```
`cells'` or `cell's` ?

vadz commented:
Personally I use `wxWindow*` because it's consistent with e.g. `wxDC&` and also because it's the only form that can really be used as `wxWindow* const` (`wxWindow *const` is plain weird IMO), but yes, you're probably right and we shouldn't use it in new wx code, I've fixed all the occurrences of it except for the one with `const`.

Also fixed the typo and changed the docs to say just "data", "cell" is inconsistent with "data of the items" used below and the latter is arguably more clear, even more verbose.

discnl commented:
> Personally I use wxWindow* because it's consistent with e.g. wxDC& and also because it's the only form that can really be used as wxWindow* const (wxWindow *const is plain weird IMO), but yes, you're probably right and we shouldn't use it in new wx code, I've fixed all the occurrences of it except for the one with const.

I'm fine with any style: no strict rules for these minor styles, or sticking to what is mostly used in wx, or having some newer preference for newly added code. It's more about knowing which one to follow. For example for the pointers the guidelines don't mention anything, at best some unrelated function declaration mentioning `const char *text`.

discnl commented on src/generic/markuptext.cpp:
```diff
@@ -217,12 +190,96 @@ class wxMarkupParserRenderOutput : public wxMarkupParserAttrOutput
+
+private:
+    wxWindow* const m_win;
+    int const m_rendererFlags;
```
`const int`? At least it's a way more common style in wx.

vadz replied:
Yes, but it's inconsistent with the lines just above and below it...

discnl replied:
I noticed that but forgot to mention it as a reason. Alrighty then :-).

vadz commented:
De facto, the existing code uses `T *` much more than `T*`, but the latter makes more sense (because of `T&` and `T* const` considerations), so I'm just not sure.

vadz commented:
Merged in f792ba45f6e062fbea699aeb43e126df1e78b605

iwbnwif commented:
This is a great enhancement, thanks!

Might it be worth populating the model in the sample differently if wxUSE_MARKUP is not set? This is because currently the sample shows the entire markup in the "attributes" column which is a bit untidy.

vadz commented:
I thought it wasn't worth it because `wxUSE_MARKUP` was going to be 1 in 99% of cases anyhow. If you explicitly disable it, I don't think it's such a big deal to see what you're missing in the sample :-)
